impl/src/main/java/org/jboss/weld/event/TransactionalObserverNotifier.java

```diff
@@ -73,8 +78,32 @@ protected TransactionalObserverNotifier(String contextId, TypeSafeObserverResolv
             // Transaction is not active - no deferred notifications
             notifySyncObservers(observers, event, metadata);
         } else {
-            for (ObserverMethod<? super T> observer : observers) {
-                deferNotification(event, metadata, observer);
+            try {
+                for (ObserverMethod<? super T> observer : observers) {
+                    deferNotification(event, metadata, observer);
+                }
+            } catch (Exception e) {
+                if (e.getCause() instanceof RollbackException || e.getCause() instanceof IllegalStateException) {
+
+                    List<ObserverMethod<? super T>> filteredObservers = new ArrayList<ObserverMethod<? super T>>();
+                    filteredObservers.addAll(observers);
+                    for (ObserverMethod<? super T> observer : observers) {
```

Comment:
I think a better approach would be to iterate over `filteredObservers` and use iterator to remove AFTER_SUCCESS observers.